drivers/dma/dma_nxp_4ch_dma.c

```diff
@@ -0,0 +1,474 @@
+ */
+static int nxp_dma_start(const struct device *dev, uint32_t channel)
+{
+	const struct nxp_dma_config *cfg = dev->config;
+	struct nxp_dma_data *data = dev->data;
+	struct nxp_dma_chan_data *chan_data = &data->chan[channel];
+	DMA_Type *dma = cfg->dma_base;
+
+	if (channel >= cfg->num_channels) {
+		return -EINVAL;
+	}
+
+	if (chan_data->busy) {
+		return -EBUSY;
+	}
+
+	if ((dma->DMA[channel].DSR_BCR & DMA_DSR_BCR_BCR_MASK) == 0U) {
+		return -EINVAL;
+	}
+
+	chan_data->busy = true;
+
+	/* Start: either allow HW request or issue software start */
+	if ((dma->DMA[channel].DCR & DMA_DCR_ERQ_MASK) == 0U) {
+		dma->DMA[channel].DCR |= DMA_DCR_START_MASK;
+	}
+
+	return 0;
+}
+
+static int nxp_dma_stop(const struct device *dev, uint32_t channel)
+{
+	const struct nxp_dma_config *cfg = dev->config;
+	struct nxp_dma_data *data = dev->data;
+	struct nxp_dma_chan_data *chan_data = &data->chan[channel];
+	DMA_Type *dma = cfg->dma_base;
+
+	if (channel >= cfg->num_channels) {
+		return -EINVAL;
+	}
+
+	/* Disable ERQ to stop further HW requests */
+	dma->DMA[channel].DCR &= ~DMA_DCR_ERQ_MASK;
+
+	/* Disable DMAMUX channel. */
+	if (cfg->dmamux_base) {
+		cfg->dmamux_base->CHCFG[channel] &= (uint8_t)~DMAMUX_CHCFG_ENBL_MASK;
+	}
+
+	/* Capture remaining BCR and then restore the remaining BCR to enable
+	 * subsequent continuation via start() without calling configure()/reload().
+	 * Note that SAR/DAR have advanced to the current position during transmission
+	 * and are intentionally not restored to enable 'resumable download'.
+	 */
+	uint32_t remain = (dma->DMA[channel].DSR_BCR & DMA_DSR_BCR_BCR_MASK) >>
+				DMA_DSR_BCR_BCR_SHIFT;
+
+	/* Clear status/error bits; note this also clears BCR to 0 per RM */
+	dma->DMA[channel].DSR_BCR |= DMA_DSR_BCR_DONE_MASK;
+
+	if (remain != 0U) {
+		dma->DMA[channel].DSR_BCR = DMA_DSR_BCR_BCR(remain);
+	}
+
+	chan_data->busy = false;
+
+	return 0;
+}
+
+static int nxp_dma_reload(const struct device *dev, uint32_t channel,
+			uint32_t src, uint32_t dst, size_t size)
+{
+	const struct nxp_dma_config *cfg = dev->config;
+	struct nxp_dma_data *data = dev->data;
+	DMA_Type *dma = cfg->dma_base;
+	struct nxp_dma_chan_data *chan_data = &data->chan[channel];
+
+	if (channel >= cfg->num_channels) {
+		return -EINVAL;
+	}
+
+	/* Alignment requirements: address aligned to transfer width */
+	if ((src % chan_data->width) != 0U || (dst % chan_data->width) != 0U) {
+		return -EINVAL;
+	}
+
+	if (chan_data->busy) {
+		return -EBUSY;
+	}
+
+	/* Configure SAR/DAR and BCR */
+	dma->DMA[channel].SAR = src;
+	dma->DMA[channel].DAR = dst;
+	dma->DMA[channel].DSR_BCR = DMA_DSR_BCR_BCR(size);
+
+	return 0;
+}
+
+static int nxp_dma_get_status(const struct device *dev, uint32_t channel,
+			struct dma_status *status)
+{
+	const struct nxp_dma_config *cfg = dev->config;
+	struct nxp_dma_data *data = dev->data;
+	DMA_Type *dma = cfg->dma_base;
+	struct nxp_dma_chan_data *chan_data = &data->chan[channel];
+
```

Comment:
Better to add NULL pointer check for status.

```suggestion
    if (status == NULL) {
        return -EINVAL;
    }
```